### ISSUE TITLE
Low zero-shot performance on Xenium breast

### ISSUE THREAD

**Issue by DhruvaRajwade** [peng-lab/phoenix#17]
Hi — thanks for releasing Phoenix. Related to #9, we're using the released TENX checkpoint as a zero-shot baseline on 3 Xenium human-breast slides (HEST-1k `NCBI783/784/785`), and we're getting lower numbers than we expected, so we'd like to confirm we're running inference and preprocessing correctly.

## What we observe

Scored per gene on the slide's measured genes, top-50 mean (zero-bin applied, counts ≤ 1 → 0), mean over the 3 slides:

| metric | value |
| --- | --- |
| Pearson, top-50 | **~0.154** |
| Spearman, top-50 | **~0.145** |

It varies a lot by slide (one slide is much lower than the others). At the single-cell level on the same data a plain linear regressor on our foundation features actually edges it out, which is what makes us think we may be preprocessing the data incorrectly for phoenix.

## How we load the model

```python
import timm
from phoenix.models.flow_simple import FlowTransformerModel, FlowTransformerConfig
from phoenix.helpers.inference import FlowPipeline

vision_model = timm.create_model(
    "vit_giant_patch14_reg4_dinov2", pretrained=False, img_size=224,
    num_classes=0, global_pool="token", init_values=1e-5, dynamic_img_size=False,
)
flow_model = FlowTransformerModel(
    FlowTransformerConfig(
        d_genes=1, d_image=1536, d_model=512, d_cross=512, n_heads=8,
        n_layers=8, qkv_bias=False, ffn_bias=False, ffn_mult=4,
        attn_drop=0.0, proj_drop=0.0, n_classes=0, cls_drop=0.1, checkpoint=False,
    ),
    vision_model=vision_model,
)
flow_model.load_state_dict(torch.load("flow_model_tenx.pth", map_location="cpu"), strict=True)   # strict load, no missing/unexpected keys
flow_model = flow_model.eval().to(device)

pipeline = FlowPipeline(model=flow_model, stats=np.load("stats_table_tenx.npz"),
                        t_0=0.0, t_1=1.0, atol=1e-1, rtol=1e-1)

gex_pred, _ = pipeline(gene_list, dataloader)   # (N, 377) — gene_list from xenium_human_multi.npy
```

## How we preprocess the H&E

We evaluate on **55 µm bins** (to compare against binned Xenium ground truth), so we run
**one 224×224 patch per bin**, centered on the bin's mean transcript centroid:

- **FOV**: 112 µm = 224 px × 0.5 µm/px. We crop
  `fov_px = round(112 µm / wsi_pixel_size_um)` at WSI level 0, then resize the crop to 224.
- **Center**: mean `(he_x, he_y)` of the bin's transcripts 
- **Transform** applied to each 224×224 patch:

```python
from torchvision.transforms import v2, InterpolationMode
transform = v2.Compose([
    v2.Resize((224, 224), InterpolationMode.BICUBIC),
    v2.CenterCrop((224, 224)),
    v2.ToTensor(),
    v2.Normalize(mean=(0.707223, 0.578729, 0.703617),
                 std =(0.211883, 0.230117, 0.177517)),
])
```

- **Ground truth**: 55 µm log1p (kNN-smoothed), restricted to each slide's measured genes.

## Questions

1. **Bin-center vs per-cell.** The paper runs per-cell (nucleus-centered) patches; we run
   one patch per 55 µm bin as a simplification. Is that the main cause of the low numbers —
   i.e. should we run per-nucleus patches and aggregate predictions to the bin? Is there a
   recommended aggregation?
2. **FOV / resolution.** Is 112 µm FOV at 0.5 µm/px → 224 px correct for the released checkpoint?
   Our WSIs aren't natively 20×; we crop at native resolution and resize to 224 — is that OK,
   or does the model expect the image already at 0.5 µm/px before the 224 resize?
3. **Normalization.** Are the mean/std above the correct values for this checkpoint? We'd like
   to confirm these are Phoenix's expected image stats (vs ImageNet or a stain-normalized set).
4. **ODE tolerances.** We use `atol = rtol = 1e-1` in `FlowPipeline`. Is that too loose for
   accurate sampling? Is there a recommended tolerance or fixed step count for evaluation?
5. **`stats_table_tenx.npz`.** Are we passing the gene-expression normalization stats to the
   pipeline correctly, and does the pipeline handle the un-normalization of the sampled genes
   internally?
6. **Vision weights.** We build the DINOv2 encoder with `pretrained=False` and expect its
   weights to come from `flow_model_tenx.pth`. Is that right, or should the DINOv2 backbone be
   loaded from a separate checkpoint?
 Any pointer on which of the above is most likely wrong would be a big help!

Thanks!

### Versions

```python

```

**Comment by DhruvaRajwade** [comment]
Hi @rushin682 @manuel-tran would you have predictions for HEST-1k slides I can use as a SOTA upper bound/reference for H&E->Gene expression? My attempts (above) at measuring performance show weird results

**Comment by rushin682** [comment]
Hi @DhruvaRajwade , in HEST-1K, are you 100% sure you're using 55um bins? As we understand, HEST-1K provides 100um bins by default not 55um. Also, HEST-1K relies on the estimated HnE pixel size to pool transcripts. Pathology scans are notorious for having variable HnE pixel sizes. As a result the the pooled transcripts don't represent the same area of pooled transcripts per patch. We believe, using this approach provides inaccuracies in comparing the signal for 55um/100um spots across multiple samples.

One option is, you could try to use our preprocessing pipeline in spatalrefinery (https://github.com/peng-lab/spatialrefinery) on the raw output bundle of NCBI samples. That way, at least we know the preprocessing issue is eliminated. TL;DR spatialrefinery uses SpatialData inherently and performs psuedo-bulking of spots at 55um and 100um consistently by using actual transcript coordinates instead of HnE shifted coordinate spaces.

Secondly, in Phoenix we have independent models one each for cell, 55um and 100um. However, we have only released the weights for the cell model since that is the highest performer. If you'd like to evaluate on HEST-1K, I'd suggest trying the 100um model weights which we can share via email, however keep in mind, it still could incur HEST-1K's preprocessing instabilities.

If your performance doesn't show improvement, we are happy to diagnose it further.

**Comment by DhruvaRajwade** [comment]
Thanks for the response! Yes, I am binning in transcript co-ord space to 55um, I'll try the spatielrefinery framework on the raw data, and contact you guys by email to get the weights for the 100um model
